mindcv/models/volo.py

```diff
@@ -0,0 +1,872 @@
+        self.k = self.kernel_size[0] * self.kernel_size[1]
+        self.c = channels
+        self.ck = self.c * self.k
+        init_weight = np.zeros((self.ck, 1, self.kernel_size[0], self.kernel_size[1]))
+        for i in range(self.ck):
+            xy = i % self.k
+            x = xy // self.kernel_size[1]
+            y = xy % self.kernel_size[1]
+            init_weight[i, 0, x, y] = 1
+        
+        # print("self.c", channels, "kernel_size", self.kernel_size, "padding", self.padding, "stride", self.stride, "dilation", self.dilation)
+        self.weight = ms.Tensor(init_weight, ms.float16)
+        self.conv_transpose2d = ops.Conv2DTranspose(self.ck, self.kernel_size,
+                                                    pad_mode="pad", pad=(self.padding[0], self.padding[0], self.padding[1], self.padding[1]),
+                                                    stride=stride, dilation=dilation, group=self.c)
+        self.reshape = ops.Reshape()
+
+    def construct(self, x: Tensor) -> Tensor:
+        b, ck, l = x.shape
+        # todo: assert is not allowed in construct, how to check the shape?
+        # assert ck == self.c * self.k
+        # assert l == self.h * self.w
+        # print("construct-b", b, "construct-ck", ck, "construct-l", l)
+        # print("self.h", self.h, "self.w", self.w)
+        x = self.reshape(x, (b, ck, self.h, self.w))
+        out = self.conv_transpose2d(x, self.weight, (b, self.c, self.output_size[0], self.output_size[1]))
+        
+        return out
+
+
+class OutlookAttention(nn.Cell):
+    """
+    Implementation of outlook attention
+    --dim: hidden dim
+    --num_heads: number of heads
+    --kernel_size: kernel size in each window for outlook attention
+    return: token features after outlook attention
+    """
+
+    def __init__(
+        self,
+        dim,
+        num_heads,
+        kernel_size=3,
+        padding=1,
+        stride=1,
+        qkv_bias=False,
+        qk_scale=None,
+        attn_drop=0.,
+        proj_drop=0.
+    ) -> None:
+        super().__init__()
+        head_dim = dim // num_heads
+        self.num_heads = num_heads
+        self.kernel_size = kernel_size
+        self.padding = padding
+        self.stride = stride
+        self.scale = qk_scale or head_dim**-0.5
+
+        self.v = nn.Dense(dim, dim, has_bias=qkv_bias)
+        self.attn = nn.Dense(dim, kernel_size**4 * num_heads)
+
+        self.attn_drop = nn.Dropout(1.0 - attn_drop)
+        self.proj = nn.Dense(dim, dim)
+        self.proj_drop = nn.Dropout(1.0 - proj_drop)
+
+        self.unfold = nn.Unfold(ksizes=[1, kernel_size, kernel_size, 1], strides=[1, stride, stride, 1],
+                                rates=[1, 1, 1, 1])
+        self.pool = nn.AvgPool2d(kernel_size=stride, stride=stride)
+        self.softmax = nn.Softmax(axis=-1)
+        self.reshape = ops.Reshape()
```

Comment:
参考：[Model Template and Guideline](https://github.com/mindspore-lab/mindcv/blob/main/docs/en/how_to_guides/write_a_new_model.md) 

When calling an operator (such as concat, reshape, mean), use the functional interface mindspore.ops.functional (such as output=ops.concat((x1, x2)) to avoid instantiating the original operator ops.Primary (such as self.Concat()) in init before calling it in construct (output=self.concat((x1, x2)).